ts/src/main/scala/ch/epfl/ts/data/StrategyParameters.scala

```diff
@@ -0,0 +1,183 @@
+ * Each parameter checks the validity of its value on instantiation.
+ * 
+ * It can be constructed with an arbitrary number of parameters.
+ */
+class StrategyParameters(params: (String, Parameter[_])*) {
+  type Key = String
+  val parameters = params.toMap
+  
+  /**
+   * Similar to what a map's `get` would do.
+   * We also perform type checking.
+   */
+  def get[T : ClassTag](key: Key): Option[T] = {
+    parameters.get(key) match {
+      case Some(p) => p.get() match {
+        case v: T => Some(v)
+        case _ => None
+      }
+      case _ => None
+    }
+  }
+  
+  /**
+   * Get the value for this key
+   * or fallback on the provided default if:
+   *   - Such a key doesn't exist
+   *   - The key exists but doesn't have the right parameter type
+   */
+  def getOrElse[T : ClassTag](key: Key, fallback: T): T =
+    get[T](key) match {
+      case Some(v) => v
+      case _ => fallback
+    }
+  
+  def getOrDefault[T : ClassTag](key: Key, parameterType: ParameterTrait[T]): T =
+    getOrElse(key, parameterType.defaultValue)
+  
+  override def toString: String = {
+    val strings = (for {
+      p <- parameters
+      key = p._1
+      paramType = p._2.toString()
+      value = p._2.get().toString()
+    } yield key + " (type " + paramType + ") = " + value).toList
+    
+    strings.reduce((a, b) => a + '\n' + b)
+  }
+}
+
+
+
+/**
+ * Represents a generic trading strategy parameter.
+ * Strategies can declare required and optional parameters.
+ * Each parameter has a "range" of valid values.
+ * @param T Parameter value type
+ */
+abstract class Parameter[T](val name: String) {
+  /**
+   * At construction, ensure that the given value is legal for this parameter.
+   */
+  assert(isValid, "Illegal value " + get() + " for strategy parameter " + companion.name)
+  
+  /** Retrieve the value for this parameter */
+  def get(): T
+  
+  /** The companion object of this parameter */
+  protected def companion: ParameterTrait[T]
+  
+  /**
+   * Whether or not this particle instance has been
+   * parameterized with a legal value.
+   */
+  def isValid: Boolean = companion.isValid(get())
+  
+  override def toString: String = get() + " (type: " + companion.name + ")"
+}
+
+
+
+/**
+ * "Static" methods that should be implemented in each concrete
+ * parameter's companion object.
+ * If the designer of a strategy wishes to change any of the fields
+ * in this trait, it is easy to extend it and override the field.
+ */
+trait ParameterTrait[T] {
+  /** Make a new instance of the associated parameter */
+  def getInstance(v: T): Parameter[T]
+  
+  /** Name of this parameter type */
+  def name: String = this.getClass.getName
+  
+  /**
+   * @return Whether the value is suitable for this parameter
+   */
+  def isValid(value: T): Boolean
+
+  /**
+   * Generate the range of valid values for this parameter.
+   * TODO: How to handle infinite ranges? Use streams?
+   */
+  def validValues: Iterable[T]
```

Comment:
I would rather have the set of candidate values for each parameter (as opposed to all the valid values).